src/python/pants/backend/visibility/rule_types_test.py

```diff
@@ -774,3 +833,120 @@ def test_file_specific_rules(rule_runner: RuleRunner) -> None:
+            "anchor-mode/proj-root/a/b/dependency/BUILD": "files(name='txt', sources=['**/*.root-b'])",
+            "anchor-mode/proj-root/a/b/origin/BUILD": "files(name='txt', sources=['**/*.root-b'])",
+            "anchor-mode/declared/a/b/dependency/BUILD": "files(name='txt', sources=['**/*.dec-b'])",
+            "anchor-mode/declared/a/b/origin/BUILD": "files(name='txt', sources=['**/*.dec-b'])",
+            "anchor-mode/invoked/a/b/dependency/BUILD": "files(name='txt', sources=['**/*.inv-b'])",
+            "anchor-mode/invoked/a/b/origin/BUILD": "files(name='txt', sources=['**/*.inv-b'])",
+            "anchor-mode/floating/a/b/dependency/BUILD": "files(name='txt', sources=['**/*.float-b'])",
+            "anchor-mode/floating/a/b/origin/BUILD": "files(name='txt', sources=['**/*.float-b'])",
+            "anchor-mode/proj-root/BUILD": dedent(
+                """
+                files(name="root", sources=["**/*.root"])
+
+                __dependencies_rules__(
+                  ("*", "//dependency/*", "!*"),
+                )
+
+                __dependents_rules__(
+                  ("*", "//origin/*", "!*"),
+                )
+                """
+            ),
+            "anchor-mode/declared/BUILD": dedent(
+                """
+                files(name="dec", sources=["**/*.dec"])
+
+                __dependencies_rules__(
+                  ("*", "/dependency/*", "!*"),
+                )
+
+                __dependents_rules__(
+                  ("*", "/origin/*", "!*"),
+                )
+                """
+            ),
+            "anchor-mode/invoked/BUILD": dedent(
+                """
+                files(name="inv", sources=["**/*.inv"])
+
+                __dependencies_rules__(
+                  ("*", "../dependency/*", "!*"),
+                )
+
+                __dependents_rules__(
+                  ("*", "../origin/*", "!*"),
+                )
+                """
+            ),
+            "anchor-mode/floating/BUILD": dedent(
+                """
+                files(name="float", sources=["**/*.float"])
+
+                __dependencies_rules__(
+                  ("*", "dependency/*", "!*"),
+                )
+
+                __dependents_rules__(
+                  ("*", "origin/*", "!*"),
+                )
+                """
+            ),
+            **{path: "" for path in files},
+        },
+    )
+    return True
+
+
+def test_relpath_for_file_targets(rule_runner: RuleRunner, setup_testbed: bool) -> None:
+    # Testing purpose:
+    #
+    # When a file is owned by a target declared in a parent directory, make sure the correct BUILD
+    # file is consulted for the rules set to apply, and with the correct relpath for the matching.
+    assert_dependency_rules(
+        rule_runner,
+        "anchor-mode/invoked/origin/file1.inv:../inv",
+        ("anchor-mode/invoked/dependency/file2.inv:../inv", DependencyRuleAction.ALLOW),
```

Comment:
sneaky use of relative rule globs, outside of the path of the owning BUILD file.